src/azul/service/manifest_service.py

```diff
@@ -2002,3 +2006,105 @@ def qualify(qualifier, column_name, index=None):
             # Join concatenated values using the joiner
             row = {k: self.padded_joiner.join(sorted(v)) if isinstance(v, set) else v for k, v in row.items()}
             bundle_tsv_writer.writerow(row)
+
+
+@attrs.frozen(kw_only=True)
+class ReplicaKeys:
+    """
+    Most replicas contain a list of the replica's hubs. However, some entities
+    (e.g. projects) have too many hubs to track within the replica document.
+    Replicas of such entities are instead retrieved by their entity ID.
+    """
+    hub_id: str
+    implicit_hub_entity_id: str
+
+    @classmethod
+    def prepare_query(cls, keys: Iterable[Self]) -> Q:
```

Comment:
This is what I call "over-extracted". There is no good reason to move this method here. It doesn't hide (or abstract away) any complicated details about replica keys. It's better left inline at the call site.